datalad_next/processors/pattern_processor.py

```diff
@@ -0,0 +1,77 @@
+"""
+ Generator to ensure that a pattern is completely contained in single chunks,
+ if it is present in the input chunks.
+"""
+
+from __future__ import annotations
+
+from typing import (
+    Generator,
+    Iterable,
+)
+
+
+def pattern_processor(iterable: Iterable,
+                      pattern: str | bytes
+                      ) -> Generator[str | bytes, None, None]:
+    """ Yield data chunks that contain a complete pattern, if it is present
+
+    This processor ensures that a given data pattern (if it exists in the data
+    chunks) is either completely contained in a chunk or not in the chunk. That
+    means the processor ensures that all data chunks have at least the length of
+    the data pattern and that they do not end with a prefix of the data pattern.
+
+    As a result, a simple `pattern in data_chunk` test is sufficient to
+    determine whether a pattern appears in the data stream.
+
+    The processor guarantees:
+
+    1. All chunks have at minimum the size of the pattern
+    2. If a complete pattern exists, it will be contained completely within a
+       single chunk, i.e. it will NOT be the case that a prefix of the pattern
+       is at the end of a chunk, and the rest of the pattern in the beginning
+       of the next chunk
+
+    The pattern might be present multiple times in a data chunk.
```

Comment:
I think this docstring would benefit from an example. What is "pattern" exactly? Is it a regex? Is it a delimiter? If the latter, why would one want to have more than one in a yielded item, rather than using splitlines?

Reply:
started with examples in `datalad_next.itertools`

Reply:
In https://github.com/datalad/datalad-next/pull/548 i moved the docs into the docstrings of the respective functions. This helps to remove redundancies (which have substantial maintenance cost).